### ISSUE TITLE
sanitize() should sanitize input directly without clone

### ISSUE THREAD

**Issue by annevk** [WICG/sanitizer-api#129]
Instead of making a clone, sanitize() should perform the low-level sanitize operation directly upon the input. If callers want that to happen on a clone, let them take that cost.

**Comment by otherdaniel** [comment]
I very much disagree and believe this is harmful, as it goes quite directly against Sanitizer API's declared goals of being [safe and easy to use](https://github.com/WICG/sanitizer-api#goals-for-the-sanitizer-api).

The first concern is that - with DOM being what it is - cloning is harder than it appears. E.g. our implementation initially had a bug where the cloned tree inadvertently was connected and would start loading images during sanitization. Shielding developers from this sort of mis-use is important.

The second (related) concern is that a DOM tree in the browser has substantial state associated with it (e.g. JS prototypes and additional members, event handlers, it might be connected or not, it might be part of a shadow DOM or not, etc.). Additionally, there is a lot of implementation-defined state like loaders (for images or scripts). Some of this state is not even query-able, at least not directly. If we sanitize in-place, the supposedly sanitized result has largely undefined state and security properties, because whatever state it initially had is simply carried on.

A meaningful security API should have meaningful guarantees about its results. For a Sanitizer, it should be a clean, non-XSS-able DOM tree in a well-defined and non-surprising state.  One can take the legalistic view that the Sanitizer API should only be about which DOM nodes are present in the tree, with any other aspect of the DOM being whatever it happens to be. But that's hardly a good API, nor can that be a secure API.

I'll note that this general theme has shown up repeatedly in Sanitizer API's development, and we have been consistent in how we address it: We have removed abusable string-to-string APIs, to address mXSS. We have also disallowed abusable configurations (e.g. a configuration that allows `<script>` is no longer possible). The mere fact that these *could* be used correctly was considered insufficient, because an abusable API is ultimately an insecure API. In all those cases we have decided that insecure usage should be avoided entirely where possible, or at least be made difficult. The same should apply here.

If callers wish to forfeit the Sanitizer's security properties in exchange for performance, they are welcome to do so by using the existing DOM APIs.

**Comment by annevk** [comment]
Could you make these concerns more concrete, for instance by providing some examples?

It also seems to me that in the majority of cases where you would invoke `sanitize()` these concerns would not apply as you'd have a tree returned from `XMLHttpRequest` or `DOMParser`. Is that not the case?

**Comment by koto** [comment]
I think cloning here is ergonomical. Given that `.sanitize` returns a value, I would expect a fresh object, and not a modified argument. The API shape doesn't suggest that it's mutating the passed tree in-place. If the expectation is to modify in-place, perhaps `void HTMLElement.sanitizeWith(sanitizer)` would be better?

> Could you make these concerns more concrete, for instance by providing some examples?

I'd expect the sanitized tree not to have event handlers bound to its elements, @otherdaniel mentioned other vectors as well. In layman's terms, I would expect the tree to be in a clean state, safe to interact with (a.k.a. "sanitized"). Without cloning, the side-effects of operating on a tree are not dependent on a sanitizer config alone, but on what happened to that tree before it was sanitized. So I need to do something on top anyway for my code to be resilient to funky inputs.

All HTML sanitizers worth their salt I've seen so far return a cloned tree by default (e.g. [DOMPurify](https://github.com/cure53/DOMPurify/blob/main/src/purify.js#L274)), or encapsulate the logic to clone to an inert tree. Not doing so has led to bypasses in the sanitizer logic - though, to be fair, that's probably a risk for a polyfill rather than native implementations.

Cloning seems strictly safer than not, and given this API is for security purposes, perhaps it's better to flip the question. What use cases have been presented that not-cloning solves?

**Comment by annevk** [comment]
Right, I agree that once we have agreement on what this API should do, the shape maybe ought to change as well.

I find it hard to evaluate the claims about event listeners without knowing the use cases and expected usage patterns for this API. So I'd appreciate more detail on where these risks could be introduced relative to the parsing step which presumably happens in isolation and without sanitization.

(Personally I don't really see the need for this API and would much rather we just ship `setHTML` and see where that takes us.)

**Comment by otherdaniel** [comment]
> Could you make these concerns more concrete, for instance by providing some examples?

I'm not sure what you're asking about. An API should have some meaningful description of what it does; that particularly applies to a security API. For a Sanitizer, I'd very much expect to obtain a safe approximation of my input, even if I pass in unsafe or surprising input (for pretty much any definition or aspect of "safe").

Real-world examples are DOMPurify (which parses internally, and optionally returns the resulting node tree or a string rendition, this guaranteeing a "clean" resulr), or the Closure library's sanitizer (same). In fact, any client-side HTML sanitizer I have looked at follows that pattern.

A real-world counter-example is crbug.com/1159750, where we inadvertently created a connected DOM tree and sanitized into it. If we - the people building the browser - can fall into that trap, I'm going to bet others might, too. The API should protect them from it.


> It also seems to me that in the majority of cases where you would invoke `sanitize()` these concerns would not apply as you'd have a tree returned from `XMLHttpRequest` or `DOMParser`. Is that not the case?

That argument misses the point. In all other cases, we have eliminated mis-usable APIs from Sanitizer, rather than hoping that concerns would not apply with common usage patterns. 

The majority of sanitizer usage I have seen is building up UI elements, usually based on some sort of template system that would construct strings and then pass them to whatever sanitizer they have. The exact workings of template systems vary widely. If you think other usage is common, can you please substantiate that and provide some examples? _"It also seems to me that in the majority of cases"_ is unfortunately not much to go by. What is your assessment based on?

**Comment by annevk** [comment]
I'm not sure how DOMPurify is comparable if it does its own parsing. There's several questions here:

1. How do you end up with a tree you end up passing to this API?
2. Can others mess with that tree in a way that is not already problematic to your application?
3. If that tree were sanitized without cloning, what new problems can this create?

I don't think any of them have been answered thus far, but I might be missing something.
